### PR TITLE
CTCP-3299: New LRN

### PR DESCRIPTION
**Pull request by Reece-Carruthers**

### CODE REVIEW

**Comment by Reece-Carruthers** [comment]
rebuild

File: app/services/DuplicateService.scala [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+      cacheConnector.post(
+        updatedUserAnswers
+      ) // TODO CTCP-3469 Will have to keep any draft declaration with same LRN, can probably handle this when the isDuplicateLrn is called in the backend
+    case None => Future.successful(false)
+  }
+
+  def isDuplicateLRN(lrn: LocalReferenceNumber)(implicit hc: HeaderCarrier): Future[Boolean] =
+    cacheConnector.isDuplicateLRN(lrn)
+
+  def apiLRNCheck(lrn: LocalReferenceNumber)(implicit hc: HeaderCarrier): Future[Boolean] =
```
Comment (jackheslop96) [inline app/services/DuplicateService.scala:49]: Can this be better named? What are we checking?

File: app/services/DuplicateService.scala [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    cacheConnector.apiLRNCheck(lrn)
+
+  def populateForm(submittedValue: Option[LocalReferenceNumber])(implicit hc: HeaderCarrier): Future[Form[LocalReferenceNumber]] =
+    submittedValue match {
+      case Some(newLocalReferenceNumber) => isDuplicateLRN(newLocalReferenceNumber).map(formProvider(_))
+      case None                          => Future.successful(formProvider(alreadyExists = false))
+    }
```
Comment (jackheslop96) [inline app/services/DuplicateService.scala:56]: Would prefer this method to return a boolean, then populate the form in the controller

File: app/controllers/NewLocalReferenceNumberController.scala [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    with I18nSupport {
+
+  private def form(alreadyExists: Boolean = false): Form[LocalReferenceNumber] = formProvider(alreadyExists)
+
+  def onPageLoad(oldLocalReferenceNumber: LocalReferenceNumber): Action[AnyContent] = identify.async {
+    implicit request =>
+      duplicateService.apiLRNCheck(oldLocalReferenceNumber).flatMap {
+        case true  => Future.successful(Ok(view(form(), oldLocalReferenceNumber)))
+        case false => Future.successful(Redirect(controllers.routes.ErrorController.badRequest())) //TODO: More generic error page?
+      }
```
Comment (jackheslop96) [inline app/controllers/NewLocalReferenceNumberController.scala:52]: If you change this to `map` you won't need the `Future.successful` in both cases

File: app/controllers/NewLocalReferenceNumberController.scala [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+      val submittedValue = form().bindFromRequest().value
+      duplicateService.populateForm(submittedValue).flatMap {
+        _.bindFromRequest()
+          .fold(
+            formWithErrors => Future.successful(BadRequest(view(formWithErrors, oldLocalReferenceNumber))),
+            newLocalReferenceNumber =>
+              duplicateService.copyUserAnswers(oldLocalReferenceNumber, newLocalReferenceNumber) flatMap {
+                case true  => Future.successful(Redirect(controllers.routes.TaskListController.onPageLoad(newLocalReferenceNumber)))
+                case false => Future.successful(Redirect(controllers.routes.ErrorController.technicalDifficulties()))
+              }
```
Comment (jackheslop96) [inline app/controllers/NewLocalReferenceNumberController.scala:66]: If you change this to `map` you won't need the `Future.successful` in both cases

File: app/connectors/CacheConnector.scala [hunk 5]
```diff
@@ -78,4 +78,25 @@ class CacheConnector @Inject() (
     }
   }
 
+  def isDuplicateLRN(lrn: LocalReferenceNumber)(implicit hc: HeaderCarrier): Future[Boolean] = {
```
Comment (jackheslop96) [inline app/connectors/CacheConnector.scala:81]: `doesDraftOrSubmissionExistForLrn` ?

File: app/forms/mappings/Formatters.scala [hunk 6]
```diff
@@ -169,6 +169,38 @@ trait Formatters {
...
+                  Left(Seq(FormError(key, invalidFormatKey)))
+                } else if (alreadyExists) {
+                  Left(Seq(FormError(key, alreadyExistsKey)))
+                } else {
+                  LocalReferenceNumber(str).map(Right.apply).getOrElse(Left(Seq(FormError(key, invalidCharactersKey))))
+                }
+              } else {
+                Left(Seq(FormError(key, lengthKey)))
+              }
+          }
```
Comment (jackheslop96) [inline app/forms/mappings/Formatters.scala:198]: Possible refactoring (although unsure if it works or not)

```
override def bind(key: String, data: Map[String, String]): Either[Seq[FormError], LocalReferenceNumber] = {
        lrnFormatter(requiredKey, lengthKey, invalidCharactersKey, invalidFormatKey)
          .bind(key, data)
          .flatMap {
            lrn =>
              if (alreadyExists) {
                Left(Seq(FormError(key, alreadyExistsKey)))
              } else {
                Right(lrn)
              }
          }
```

File: test/generators/ModelGenerators.scala [hunk 7]
```diff
@@ -25,6 +25,13 @@ import uk.gov.hmrc.http.HttpVerbs._
 trait ModelGenerators {
   self: Generators =>
 
+  implicit lazy val arbitrarySubmissionState: Arbitrary[SubmissionState] = Arbitrary {
+    val validStates = Seq("notSubmitted", "submitted", "rejectedPendingChanges", "rejectedAndResubmitted")
```
Comment (jackheslop96) [inline test/generators/ModelGenerators.scala:29]: Could use the toString definitions here, just so you're not doubling up on static strings

**Review by jackheslop96 — APPROVED** [review]

**Pull request merged**